### ISSUE TITLE
OpenAi prompt/completion support

### ISSUE THREAD

**Issue by rlbns** [langchain4j/langchain4j#38]
**Is your feature request related to a problem? Please describe.**
I want to provide context to a model that uses the prompt/completion pattern as follows:

{"prompt": "burger -->", "completion": " edible"}
{"prompt": "paper towels -->", "completion": " inedible"}
{"prompt": "vino -->", "completion": " edible"}
{"prompt": "bananas -->", "completion": " edible"}
{"prompt": "dog toy -->", "completion": " inedible"}

ref: https://docs.google.com/document/d/1rqj7dkuvl7Byd5KQPUJRxc19BJt8wo0yHNwK84KfU3Q/edit

**Describe the solution you'd like**
A dedicated class in the model.openai package would be great. Some instructions on how to do this or an example would help in the short term.

**Describe alternatives you've considered**
I've looked at all the classes in langchain4j and can't find anything related to this. Perhaps I've simply missed it. I've also searched a lot on Google.

This library has some support for this pattern:
https://github.com/TheoKanning/openai-java

It uses a FineTuneRequest class which points to a JSON Lines document uploaded to openai (note the above example is in that format). My fallback is to use this library together with langchain4j, but I would prefer to use langchain4j for everything if possible.

**Additional context**
Great job on this library. It is well written and quite easy to understand in most classes.

**Comment by langchain4j** [comment]
@rlbns Thank you for kind words, it means a lot to us :)

Regarding your use case:
The [reference](https://docs.google.com/document/d/1rqj7dkuvl7Byd5KQPUJRxc19BJt8wo0yHNwK84KfU3Q/edit) you provided seems to be an outdated one and explains how to **fine-tune** old davinci models which [will be deprecated soon](https://openai.com/blog/gpt-4-api-general-availability). OpenAI recommends using chat completion with gpt-3.5-turbo instead of davinci models. They are better and much (20-30x) cheaper. In general, they will continue development and support of **chat completion** models/endpoints, and will no longer support older **completion** models/endpoints (which still can be fine-tuned, but this will end soon) because chat API proved to be very intuitive and versatile.

From what I understood, your aim is not to **fine-tune** your own model (based on their base model), but to do an **in-context learning** by providing a few examples to a model on how it should respond. For this you don't need FineTuneRequest or similar.

At the moment I see a few options:
1. In case your use-case is similar to the abovementioned "edible/inedible" classification, gpt-3.5-turbo/gpt-4 will most probably be able to figure it out on it's own and you dont need to provide a list of examples. In this case you can use "AI Services" for classification like this:
```java
interface Classifier {

    @UserMessage("Is {{it}} edible?") // "it" can be used to refer to the only argument in the method
    boolean isEdible(String stuff);
}

Classifier classifier = AiServices.create(Classifier.class, chatLanguageModel);
System.out.println(classifier.isEdible("burger")); // true
System.out.println(classifier.isEdible("paper towels")); // false
```
You can also use enums for this (e.g. EDIBLE, INEDIBLE, NOT_SURE). You can see an example with Sentiment here: https://github.com/langchain4j/langchain4j-examples/blob/main/other-examples/src/main/java/OtherServiceExamples.java
For this example, it seems to work quite good.

2. In case your use case is more sophisticated than edible/inedible and you want to provide a few examples, you can extend the above example like this:
```java
interface Classifier {

    @UserMessage({
                    "Is {{it}} edible?",
                    "Use the following logic:",
                    "burger -> true",
                    "paper towels -> false",
                    "vino -> true",
                    "bananas -> true",
                    "dog toy -> false"
            })
    boolean isEdible(String stuff);
}
```

3. You can also use OpenAiChatModel to do a "real" in-context learning in the following way:
```java
AiMessage aiMessage = chatLanguageModel.sendMessages(
                userMessage("burger"), aiMessage("edible"),
                userMessage("paper towels"), aiMessage("inedible"),
                userMessage("vino"), aiMessage("edible"),
                userMessage("bananas"), aiMessage("edible"),
                userMessage("dog toy")
);

 System.out.println(aiMessage.text()); // inedible
```

If this will not work for your use case, we can discuss how to extend this lib to support your use case.
Hope this helps. Please let me know if you have questions.

**Comment by langchain4j** [comment]
@rlbns it would be awesome if you could provide your feedback on what you **don't** like about the library and which parts are not clear/intuitive. It will help us a lot to improve it. Thank you!

**Comment by rlbns** [comment]
Thanks so much for the detailed reply.

My use case is in fact more complicated. I want to break complex reports into paragraphs and show the AI how to convert each paragraph into data-driven XSLT instructions. For example:

"In a clinical interview with Harry, he reported that ..."

can be encoded in XSLT as:

"\<fo:block>In a clinical interview with <xsl:call-template name='Client'/>, <xsl:call-template name='he-she'/> reported that ...</fo:block>"

This will allow us to send existing reports to the AI that people have in Word (paragraph by paragraph), and get back dynamic XSLT content that we can merge into one or more XSLT documents. The result can then be paired with data gathered by forms (e.g., name, gender) to generate a draft .docx document. The above example gets more complicated when you address xsl:if, xsl:choose/when and xsl:value-of but I'm guessing the AI can handle that if the training is good enough.

Our motivation is to give medical practitioners clear instructions regarding how content will be automated, which they can review and customize for their needs. In our experience, while you can ask AI to generate the report itself, it can sometimes create spectacular errors. Someone who is responsible for a client's health will likely appreciate inspecting every rule used to automate their specialized content.

Put another way, use black box logic (AI magic) to generate white box logic (XSLT rules).

So I don't think Classifier will help because the output is more complex. This is more similar to translation between english and spanish I suppose. We can't rely on GPT to already know this domain because the result contains calls to XSLT utility templates (and data fields) that we have created.

Do you think that "real" in-context learning will work this way, where each userMessage and aiMessage is one or more paragraphs long? There will be dozens or hundreds of documents to train on, some of which are 20+ pages long (thousands of paragraphs). This obviously blows past rate limits for a single call to the AI.

My goal is to use a Pinecone vector database to store both parts of the prompt/response. I've got code that uses PineconeEmbeddingStore to store OpenAI embeddings in the database. However, I'm not sure how to leverage that for in-context learning. I assume it would be more efficient / cheaper to communicate with the AI using embeddings, but maybe that is not true.

I've seen lots of references to vector databases providing long-term memory for AI models. Our documents won't change much over time, it's just that we'll need lots of examples for each subject/report domain. Do you think that kind of approach is practical for this project?

**Comment by langchain4j** [comment]
@rlbns This is a very interesting use case. Do I understand correctly that you want to take existing medical reports and create templates out ouf them with AI by identifying variable parts (e.g. name, gender, etc)? This is not my business, but I can't help but think that this might violate some data privacy rules. Please make sure to check [OpenAI policies](https://openai.com/policies) because if you legally can't send patient data to OpenAI this will change a lot (you will have to use local models such as Falcom/Llama2/MPT or at least anonymize the report before sending it to OpenAI).

Do you already have a complete list of entities/terms that should be converted into placeholders? Like:
- Name
- Surname
- Gender
- Age
- etc

This would help if you could provide more information here (anonymized report and expected xslt).

Regarding in-context learning, I would give it a try. You can try it out using [OpenAI playground](https://platform.openai.com/playground) where you can define a sequence of examples (example user message -> expected AI response) and then provide your real user message. You can also try specifying system message where you can provide more context and define the behaviour of AI in imperative way (e.g. Your goal is to convert user message into XSLT template. You must convert each name, gender, etc into ...). If the list of things that should be replaced with placeholders is not huge, this might work. But if you have a lot of very specific things, you will have to provide examples for each of them into the request to AI and this will increase the cost and latency a lot.

You can also look into NER (Named Entity Recognition), I think it might help as well. Please check SpaCy, Stanford Named Entity Recognizer, NLTK (Natural Language Toolkit).

If both options above will not work, then I would look into the direction of fine-tuning. But here you will need to gather large amounts of training data (example report -> expected xslt) and find an LLM that will be able to learn this. As I have mentioned, OpenAI is going to depredate davinci family which you can use for fine-tuning, but AFAIK they did not yet provide a replacement:

> Developers wishing to continue using their fine-tuned models beyond January 4, 2024 will need to fine-tune replacements atop the new base GPT-3 models (ada-002, babbage-002, curie-002, davinci-002), or newer models (gpt-3.5-turbo, gpt-4). Once this feature is available later this year, we will give priority access to GPT-3.5 Turbo and GPT-4 fine-tuning to users who previously fine-tuned older models. We acknowledge that migrating off of models that are fine-tuned on your own data is challenging. We will be providing support to users who previously fine-tuned models to make this transition as smooth as possible.

Regarding Pinecone/vector stores, I am not sure I understand how you want to use it.
Vector stores are very helpful for semantic search, when you have lots of information and you need to find relevant pieces by semantic meaning. But this is fuzyy, not a precise technique like searching SQL DB where you know what you are looking for and guaranteed to get it back if it exists in the DB.

Regarding communicating with LLMs using embeddings, I am not sure this is possible at this point. Major LLM providers offer text-in text-out APIs and you will loose a lot of details when you convert text into embedding: it is mostly irreversible process where you try to fit a variable length text into a fixed-size vector. Basically this is a compression with the loss of data. So while sending embeddings to LLM (in theory) can speedup transmission, you will loose on the precision side.

Vector stores are often used to store knowledgebases and provide relevant context to the model. For example, you want to have a chatbot that is knowledgable in the medical domain. You take a lot of medical literature, embed it and put it into the vector store. Then when end user talks to the bot, on each his question you can retrieve relevant pieces of knowledgebase and provide it as a context to the model so that it can utilize this factual information for answering. This reduces hallucinations a lot. E.g. user asks "I have a headache, what can I do?" and vector store will retrieve the most relevant paragraphs/chunks of knowledgebase about headaches.

Hope this helps.

**Comment by langchain4j** [comment]
@rlbns Regarding in-context learning, you can also try it out directly with LangChain4j using `chatLanguageModel.sendMessages(...)`. But it might be easier to do it in OpenAI playground UI where it is probably more intuitive.

**Comment by rlbns** [comment]
Thanks again for all the great information.

Yes, all reports have anonymized data. We are leveraging a manual process already in place. Our reports range from simple (name, gender, etc.) to complex which have hundreds of data points used to decide which fragment of text to insert.

SpaCy and NLTK look really interesting - I appreciate the links. NER is probably not a good fit, because I hope to teach the AI about conditional fragments (if, choose/when) so it goes beyond the scope of named entities. I've worked with this before and it definitely has limits.

I spent some time on the playground. I found that davinci was way better than the other legacy models on a simple version of this task (no custom training). It was even a little better than GPT-3.5 Turbo, which was also pretty good.

I also had some success in chat mode giving a detailed prompt. This was encouraging, but clearly won't scale too well. 

I was also able to fine-tune davinci using some simple training data. It takes a long time to execute a fine-tune job these days (hours waiting in a queue I suppose). The results were surprising. My first attempt ruined davinci and all it returned was gibberish - I even got back chinese characters, no kidding. My second attempt included more imperative instructions and was better, but it still resulted in answers that were only partly correct. Setting temperature to 0 helped, but I'm still getting back a lot of strange content in addition to the correct answer,

I think I understand vector stores better. They work great for semantic search, where the database itself can execute search queries and the results can inform a LLM. However, you are right that this does not line up with what I'm trying to do.

While I could try to improve the training data for davinci, I think I'll stick a pin in it until we see how the newer models support training in the fall. They will probably be smarter in other ways as well. I'm also worried that I will bias our training data to deal with davinci in its current form, when we know it will be a GPT-* model involved in a few months.

Thanks so much for your help. I will come at this again in a few months and may post another issue to ask for advice and give you an update. I hope you do support uploading files and fine-tune requests when the new way of doing this is revealed by OpenAI. I prefer java over python for many reasons, so being able to call the OpenAI API entirely from java is very valuable to me.

As for suggestions for langchain4j, javadocs would be helpful. Even though you don't have comments, it can help new developers develop a mental map of your packages (e.g., one click to investigate a method return type). Your consistent use of the builder pattern is great, you include some great utilities (splitters, document loading, etc.), and I found the code itself is logical and easy to read. Thanks for making this open source.

**Comment by langchain4j** [comment]
@rlbns That's a very interesting experience! Did you try gpt-4? In my experience it is **way** better than davinci3/gpt3.5. Especially when you provide very detailed instructions in the system prompt, it can do wonders. If you still have your prompts somewhere, I would recommend trying it out. It is much more expensive than 3.5, but I believe is is still way cheaper than manual labor.
BTW chat completion can be used as a normal completion as well, you don't have to continue the conversation. You can just provide a system message + user message, then read response and drop the conversation.

I will try to add fine-tuning, but can't guarantee anything. This library was not intended to train the models, rather to use them. I have a huge backlog at the moment and hopefully get more contributors. For fine-tuning you can use openai-java library, it should be fine since training and using models are anyways two different processes, so you can use multiple libraries for that.

Thank you for suggestions, I will try to add that ASAP.
And again thank you so much for kind words, this helps a lot!

Please fel free to open issues in the future, I will be happy to help.
For the moment I am closing this one and wish you a good luck in your LLM journey :)

**Comment by langchain4j** [comment]
@rlbns some news on fine-tuning: https://twitter.com/OfficialLoganK/status/1686749315165835264?t=nue7B5WaePf9ojyly6yUZg&s=09
